File: ios/Runner/YuvConverter.swift
import Flutter
import UIKit

class YuvConverter {
    static let CHANNEL = "bowling_diary/yuv_converter"

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: CHANNEL, binaryMessenger: registrar.messenger())
        channel.setMethodCallHandler { call, result in
            if call.method == "convert" {
                guard let args = call.arguments as? [String: Any],
                      let width = args["width"] as? Int,
                      let height = args["height"] as? Int,
                      let yData = (args["yPlane"] as? FlutterStandardTypedData)?.data,
                      let uData = (args["uPlane"] as? FlutterStandardTypedData)?.data,
                      let vData = (args["vPlane"] as? FlutterStandardTypedData)?.data,
                      let yRowStride = args["yRowStride"] as? Int,
                      let uvRowStride = args["uvRowStride"] as? Int,
                      let uvPixelStride = args["uvPixelStride"] as? Int
                else {
                    result(FlutterError(code: "BAD_ARGS", message: "missing args", details: nil))
                    return
                }
                let rgba = convertYuvToRgba(
                    width: width, height: height,
                    y: yData, u: uData, v: vData,
                    yRowStride: yRowStride, uvRowStride: uvRowStride, uvPixelStride: uvPixelStride
                )
                result(["rgba": FlutterStandardTypedData(bytes: rgba)])
            } else {
                result(FlutterMethodNotImplemented)
            }
        }
    }

    // YUV420 → RGBA8888 변환. 카메라 시점 회전 처리는 호출자 책임.
    private static func convertYuvToRgba(
        width: Int, height: Int,
        y: Data, u: Data, v: Data,
        yRowStride: Int, uvRowStride: Int, uvPixelStride: Int
    ) -> Data {
        var rgba = Data(count: width * height * 4)
        rgba.withUnsafeMutableBytes { (rgbaPtr: UnsafeMutableRawBufferPointer) in
            y.withUnsafeBytes { (yPtr: UnsafeRawBufferPointer) in
                u.withUnsafeBytes { (uPtr: UnsafeRawBufferPointer) in
                    v.withUnsafeBytes { (vPtr: UnsafeRawBufferPointer) in
                        for j in 0..<height {
                            for i in 0..<width {
                                let yIdx = j * yRowStride + i
                                let uvIdx = (j / 2) * uvRowStride + (i / 2) * uvPixelStride
                                let Y = Int(yPtr[yIdx])
                                let U = Int(uPtr[uvIdx]) - 128
                                let V = Int(vPtr[uvIdx]) - 128
                                let r = max(0, min(255, Y + Int(Double(V) * 1.402)))
                                let g = max(0, min(255, Y + Int(Double(U) * -0.344 + Double(V) * -0.714)))
                                let b = max(0, min(255, Y + Int(Double(U) * 1.772)))
                                let out = (j * width + i) * 4
                                rgbaPtr[out]     = UInt8(r)
                                rgbaPtr[out + 1] = UInt8(g)
                                rgbaPtr[out + 2] = UInt8(b)
                                rgbaPtr[out + 3] = 255
                            }
                        }
                    }
                }
            }
        }
        return rgba
    }
}

File: android/app/src/main/kotlin/com/devpark/bowling_diary/YuvConverter.kt
package com.devpark.bowling_diary

import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

class YuvConverter {
    companion object {
        const val CHANNEL = "bowling_diary/yuv_converter"

        fun register(engine: FlutterEngine) {
            MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL)
                .setMethodCallHandler { call, result ->
                    when (call.method) {
                        "convert" -> {
                            try {
                                val args = call.arguments as Map<String, Any>
                                val width = args["width"] as Int
                                val height = args["height"] as Int
                                val yPlane = args["yPlane"] as ByteArray
                                val uPlane = args["uPlane"] as ByteArray
                                val vPlane = args["vPlane"] as ByteArray
                                val yRowStride = args["yRowStride"] as Int
                                val uvRowStride = args["uvRowStride"] as Int
                                val uvPixelStride = args["uvPixelStride"] as Int

                                val rgba = convertYuvToRgba(
                                    width, height,
                                    yPlane, uPlane, vPlane,
                                    yRowStride, uvRowStride, uvPixelStride,
                                )
                                result.success(mapOf("rgba" to rgba))
                            } catch (e: Exception) {
                                result.error("CONVERT_FAIL", e.message, null)
                            }
                        }
                        else -> result.notImplemented()
                    }
                }
        }

        // YUV420 → RGBA8888 변환. 카메라 시점 회전 처리는 호출자 책임.
        private fun convertYuvToRgba(
            width: Int, height: Int,
            y: ByteArray, u: ByteArray, v: ByteArray,
            yRowStride: Int, uvRowStride: Int, uvPixelStride: Int,
        ): ByteArray {
            val rgba = ByteArray(width * height * 4)
            for (j in 0 until height) {
                for (i in 0 until width) {
                    val yIdx = j * yRowStride + i
                    val uvIdx = (j / 2) * uvRowStride + (i / 2) * uvPixelStride
                    val Y = (y[yIdx].toInt() and 0xFF)
                    val U = (u[uvIdx].toInt() and 0xFF) - 128
                    val V = (v[uvIdx].toInt() and 0xFF) - 128

                    val r = (Y + 1.402 * V).toInt().coerceIn(0, 255)
                    val g = (Y - 0.344 * U - 0.714 * V).toInt().coerceIn(0, 255)
                    val b = (Y + 1.772 * U).toInt().coerceIn(0, 255)

                    val out = (j * width + i) * 4
                    rgba[out]     = r.toByte()
                    rgba[out + 1] = g.toByte()
                    rgba[out + 2] = b.toByte()
                    rgba[out + 3] = 0xFF.toByte()
                }
            }
            return rgba
        }
    }
}
